ast_tools/transformers/loop_unroller.py

```diff
@@ -0,0 +1,85 @@
+import ast
+from copy import deepcopy
+import astor
+
+
+class UnrollException(Exception):
+    def __init__(self, execption):
+        message = f"Unable to evaluate `unroll_range` object during " \
+            f"unrolling pass, got exception {execption}"
+        super().__init__(message)
+
+
+def is_call(node):
+    return isinstance(node, ast.Call)
+
+
+def is_name(node):
+    return isinstance(node, ast.Name)
+
+
+def replace_symbols(tree, symbol_table):
+    class SymbolReplacer(ast.NodeTransformer):
+        def visit_Name(self, node):
+            if node.id in symbol_table:
+                return deepcopy(symbol_table[node.id])
+            return node
+    return SymbolReplacer().visit(tree)
+
+
+def unroll_for_loops(tree, env):
+    class Unroller(ast.NodeTransformer):
+        def flat_visit(self, body):
+            new_body = []
+            for statement in body:
+                result = self.visit(statement)
+                if isinstance(result, list):
+                    new_body.extend(result)
+                else:
+                    new_body.append(result)
+            return new_body
+
+        def visit(self, node):
+            """
+            For nodes with a `body` attribute, we explicitly traverse them and
+            flatten any lists that are returned. This allows visitors to return
+            more than one node.
+            """
+            if not hasattr(node, 'body') or isinstance(node, ast.IfExp):
+                return super().visit(node)
+            node.body = self.flat_visit(node.body)
+            if hasattr(node, 'orelse'):
+                node.orelse = self.flat_visit(node.orelse)
+            return super().visit(node)
```

Comment:
This is entirely unnecessary.  
https://docs.python.org/3/library/ast.html#ast.NodeTransformer
> For nodes that were part of a collection of statements (that applies to all statement nodes), the visitor may also return a list of nodes rather than just a single node.

Reply:
https://github.com/python/cpython/blob/3.8/Lib/ast.py#L440

Reply:
Good catch, I can't remember why I had this code, maybe it was related to implementing a specific visitor (e.g. if I override visit_For, I have to flatten the children, but in this case it's not needed)